src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseCodeReviewFeedbackService.java

```diff
@@ -0,0 +1,239 @@
+            automaticResult.setCompletionDate(ZonedDateTime.now());
+
+            this.resultService.storeFeedbackInResult(automaticResult, feedbacks, true);
+
+            this.programmingMessagingService.notifyUserAboutNewResult(automaticResult, participation);
+        }
+        catch (Exception e) {
+            log.error("Could not generate feedback", e);
+            automaticResult.setSuccessful(false);
+            automaticResult.setCompletionDate(ZonedDateTime.now());
+            this.resultRepository.save(automaticResult);
+            this.programmingMessagingService.notifyUserAboutNewResult(automaticResult, participation);
+        }
+        finally {
+            unlockRepository(participation, programmingExercise);
+        }
+    }
+
+    /**
+     * Sets an individual due date for a participation, locks the repository,
+     * and invalidates previous results to prepare for new feedback.
+     *
+     * @param participation       the programming exercise student participation.
+     * @param programmingExercise the associated programming exercise.
+     */
+    private ProgrammingExerciseStudentParticipation setIndividualDueDateAndLockRepository(ProgrammingExerciseStudentParticipation participation,
+            ProgrammingExercise programmingExercise, boolean invalidatePreviousResults) {
+        // The participations due date is a flag showing that a feedback request is sent
+        participation.setIndividualDueDate(now());
+
+        participation = programmingExerciseStudentParticipationRepository.save(participation);
+        // Circumvent lazy loading after save
+        participation.setParticipant(participation.getParticipant());
+        programmingExerciseParticipationService.lockStudentRepositoryAndParticipation(programmingExercise, participation);
+
+        if (invalidatePreviousResults) {
+            var participationResults = participation.getResults();
+            participationResults.forEach(participationResult -> participationResult.setRated(false));
+            this.resultRepository.saveAll(participationResults);
+        }
+
+        return participation;
+    }
+
+    /**
+     * Removes the individual due date for a participation. If the due date for an exercise is empty or is in the future, unlocks the repository,
+     *
+     * @param participation       the programming exercise student participation.
+     * @param programmingExercise the associated programming exercise.
+     */
+    private void unlockRepository(ProgrammingExerciseStudentParticipation participation, ProgrammingExercise programmingExercise) {
+        if (programmingExercise.getDueDate() == null || now().isBefore(programmingExercise.getDueDate())) {
+            programmingExerciseParticipationService.unlockStudentRepositoryAndParticipation(participation);
+            participation.setIndividualDueDate(null);
+            this.programmingExerciseStudentParticipationRepository.save(participation);
+        }
+    }
```

Comment:
I see 3 smaller problems with this:
- If there was an individual due date set it is deleted after this whole thing ran through
- There is no explanation why it is required to do this. I assume that this may mess with the Athena processing, but I am sure there is a better way to do this (maybe by fixing the Athena operation to a specific commit hash?)
- Is the user properly informed about this?

Reply:
1. Regarding the first issue, yes, the due date is indeed removed (this was also the case with tutor feedback requests).

2. Currently, this action is taken to ensure Athena can fetch a consistent state of the repository. Currently, Athena doesn't support commit hashes. 

3. The user is presented with an appropriate message and can choose whether to continue.